hazelcast/src/main/java/com/hazelcast/jet/impl/connector/ReadMapOrCacheP.java

```diff
@@ -472,27 +480,96 @@ public Object toObject(@Nonnull Entry<Data, Data> dataEntry) {
     }
 
     static class LocalMapReader
-            extends Reader<InternalCompletableFuture<MapEntriesWithCursor>, MapEntriesWithCursor, Entry<Data, Data>> {
+            extends Reader<CompletableFuture<MapEntriesWithCursor>, MapEntriesWithCursor, Entry<Data, Data>> {
 
         private final MapProxyImpl mapProxyImpl;
+        private final MapServiceContext mapServiceContext;
+        private final NodeEngineImpl nodeEngine;
+        private final int retryDelay = 100;
+
 
         LocalMapReader(@Nonnull HazelcastInstance hzInstance,
                        @Nonnull InternalSerializationService serializationService,
                        @Nonnull String mapName) {
             super(mapName,
                     AbstractCursor::getIterationPointers,
                     AbstractCursor::getBatch);
-            this.mapProxyImpl = (MapProxyImpl) hzInstance.getMap(mapName);
             this.serializationService = serializationService;
+            this.mapProxyImpl = (MapProxyImpl) hzInstance.getMap(mapName);
+            nodeEngine = getNodeEngine(hzInstance);
+            MapService service = nodeEngine.getService(MapService.SERVICE_NAME);
+            mapServiceContext = service.getMapServiceContext();
         }
 
         @Nonnull @Override
-        public InternalCompletableFuture<MapEntriesWithCursor> readBatch(int partitionId, IterationPointer[] pointers) {
+        public CompletableFuture<MapEntriesWithCursor> readBatch(int partitionId, IterationPointer[] pointers) {
+            boolean isHD = mapProxyImpl.getMapConfig().getInMemoryFormat().equals(InMemoryFormat.NATIVE);
+            if (isHD)
+                return readWithOperationService(partitionId, pointers);
+
+            CompletableFuture<MapEntriesWithCursor> f = new CompletableFuture<>();
+            read0(f, partitionId, pointers);
+            return f;
+        }
+
+        private void read0(CompletableFuture<MapEntriesWithCursor> f, int partitionId, IterationPointer[] pointers) {
+            try {
+                boolean isOwned = mapServiceContext.getOrInitCachedMemberPartitions().contains(partitionId);
+                int migrationStamp = mapServiceContext.getService().getMigrationStamp();
```

Comment:
This should go after `if (isOwned)`, it's not applicable to reading using an operation